plugin-server/src/cdp/redis.ts

```diff
@@ -0,0 +1,152 @@
+// NOTE: PostIngestionEvent is our context event - it should never be sent directly to an output, but rather transformed into a lightweight schema
+
+import { captureException } from '@sentry/node'
+import { createPool } from 'generic-pool'
+import { Pipeline, Redis } from 'ioredis'
+
+import { PluginsServerConfig } from '../types'
+import { timeoutGuard } from '../utils/db/utils'
+import { status } from '../utils/status'
+import { createRedisClient } from '../utils/utils'
+
+type WithCheckRateLimit<T> = {
+    checkRateLimit: (key: string, now: number, cost: number, poolMax: number, fillRate: number, expiry: number) => T
+}
+
+export type CdpRedisClientPipeline = Pipeline & WithCheckRateLimit<number>
+
+export type CdpRedisClient = Omit<Redis, 'pipeline'> &
+    WithCheckRateLimit<Promise<number>> & {
+        pipeline: () => CdpRedisClientPipeline
+    }
+
+export type CdpRedisOptions = {
+    name: string
+    timeout?: number
+    failOpen?: boolean
+}
+
+export type CdpRedis = {
+    useClient: <T>(options: CdpRedisOptions, callback: (client: CdpRedisClient) => Promise<T>) => Promise<T | null>
+    usePipeline: (
+        options: CdpRedisOptions,
+        callback: (pipeline: CdpRedisClientPipeline) => void
+    ) => Promise<Array<[Error | null, any]> | null>
+}
+
+// NOTE: We ideally would have this in a file but the current build step doesn't handle anything other than .ts files
+const LUA_TOKEN_BUCKET = `
```

Comment:
Seeing this surprised me, but I see it was mostly (?) code movement.

Reply:
Yeah there was some context from the previous PR where @oliverb123 made the point that it would be better if we only initialised it once. Now that there is our own redis client management here I can safely do exactly that hence the code movement.